app/models/namespace.rb

```diff
@@ -225,4 +225,54 @@ def validate_nesting_level
   def restore_routes
     Route.restore(Route.only_deleted.find_by(source_id: id).id, recursive: true)
   end
+
+  # The add and subtract_from_metadata_summary count functions are used by all update_metadata_summary functions
+  # (update_service, namespace_transfer, namespace_deletion, sample_transfer, sample_deletion)
+  # to update parental namespace metadata_summary attributes.
+  #
+  # When a sample is transferred, deleted, or updated, namespaces are the non-user parental namespaces, metadata is
+  # equal to sample.metadata, and update_by_one is assigned true. Because we are passing actual metadata values
+  # (ie: sample.metadata = {metadatafield1: metadatavalue1}), we assign true to update_by_one and value = 1 when
+  # calling these functions because we will update the samples' parents by a count of 1, regardless of what the actual
+  # metadata value is (ie: sample.metadata above with a value of metadatavalue1 will result in
+  # metadata_summary['metadatafield1'] += 1 for all parental namespaces).
+  #
+  # When namespaces are transferred or deleted, namespaces are the non-user parental namespaces, metadata is equal to
+  # the namespace.metadata_summary, and update_by_one is false. With namespaces, we will pass metadata_summary
+  # (ie: project.namespace.metadata_summary = {metadatafield1: 5, metadatafield2: 10}), and we can iterate over the
+  # summary and utilize the hash values to update all parental namespaces, so setting value = 1 is unnecessary and
+  # we assign false to update_by_one (ie: all parental namespaces will have their metadata_summary updated by
+  # 5 for metadatafield1 and 10 for metadatafield2).
+  #
+  # Subtract will be called for namespace_transfer (original parents), namespace_deletion,
+  # sample_transfer (original parents), sample_deletion, and update_service when metadata is deleted from a sample.
+  def subtract_from_metadata_summary_count(namespaces, metadata, update_by_one)
+    namespaces.each do |namespace|
```

Comment:
**nitpick (blocking):** Can you add a `namespace.with_lock do` block inside so that a namespace is locked for update while this is happening so that we don't have conflicts from simultaneous updates. https://guides.rubyonrails.org/active_record_querying.html#pessimistic-locking

Reply:
Added in c945fbabc9577fb847d931c897f6c42ed635763e